libc/mem/unveil.c

```diff
@@ -135,7 +132,49 @@ static int sys_unveil_linux(const char *path, const char *permissions) {
 }
 
 /**
- * Unveil parts of a restricted filesystem view.
+ * Restricts filesystem operations, e.g.
+ *
+ *    unveil("/etc", "r");
+ *
+ * Unveiling restricts the visibility of the filesystem to a set of allowed
+ * paths with specific operations. This system call is supported natively on
+ * OpenBSD and polyfilled on Linux using the Landlock LSM[1].
+ *
+ * On OpenBSD, accessing paths outside of the allowed set raises ENOENT, and
+ * accessing ones with incorrect permissions raises EACCES. On Linux, both these
+ * cases raise EACCES.
+ *
+ * Using unveil is irreversible. On OpenBSD, the first call immediately enforces
+ * the filesystem visibilty, and existing paths can only be updated with equal
+ * or lesser permissions. Filesystem operations that try to access invisible
+ * paths will raise ENOENT, and operations without the correct permissions raise
+ * EACCES. Unveiling can be disabled by either passing two NULL arguments or by
+ * calling pledge() without the "unveil" promise.
```

Comment:
It might be worth mentioning that unveil() on Linux by itself does not restrict `open(O_PATH)`, `fstat()`, `stat()`, and in access to file system metadata it seems.

Reply:
Yes, unfortunately. See [this warning](https://docs.kernel.org/userspace-api/landlock.html#filesystem-flags) in the kernel docs:

> It is currently not possible to restrict some file-related actions accessible through these syscall families: chdir(2), truncate(2), stat(2), flock(2), chmod(2), chown(2), setxattr(2), utime(2), ioctl(2), fcntl(2), access(2). Future Landlock evolutions will enable to restrict them.